KAFKA-6987: Add KafkaFuture.toCompletionStage()

tombentley opened the pull request:
See [KAFKA-6987](https://issues.apache.org/jira/browse/KAFKA-6987) and [KIP-707](https://cwiki.apache.org/confluence/display/KAFKA/KIP-707%3A+The+future+of+KafkaFuture)

tombentley committed:
KAFKA-9687: Improve KafkaFutureTest

Improve the test prior to reimplementing KafkaFutureImpl using
CompletableFuture.

tombentley committed:
KAFKA-9687: Reimplement KafkaFutureImpl using a CompleteableFuture

* Removed public copyWith(), which was added accidentally [1].
* Removed protected addWaiter(), since it could be implemented using whenComplete()
* Didn't remove InterruptedException from throws clause of KafkaFutureImpl#getNow()
  even though it's not thrown.

[1]: https://github.com/apache/kafka/pull/5131#issuecomment-403115339

tombentley commented:
@cmccabe @ijuma could one of you take a look at this? Or @dajac perhaps?

Thus far I've stuck strictly to reimplementing without breaking the existing `KafkaFuture` API. The logical next step would be to provide an accessor to the `CompletableFuture` as mentioned [previously](https://github.com/apache/kafka/pull/5131#issuecomment-395136834).

tombentley commented:
Rebased for conflict. @chia7712 maybe you could take a look if others are too busy?

chia7712 commented:
I like this idea. We make some efforts to convert ```KafkaFuture``` to java ```CompletableFuture``` in our projects. It would be better to support ```CompletableFuture``` officially. 

However, this change can make a great impact on all producer users (bugs/performance/compatibility) so I feel it needs a KIP. (There is an KIP https://cwiki.apache.org/confluence/display/KAFKA/KIP-XXX%3A+Return+CompletableFuture+from+KafkaProducer.send)

tombentley commented:
@chia7712 to be clear, are you saying this reimplementation needs a KIP? Imho this PR as it stands should be OK without one, but of course any change to the API would need one. I was hoping to get this merged first. I'm happy to open a KIP for the API change though. I think everyone agrees that better interop between `KafkaFuture` and `CompletableFuture` would be a good thing.

chia7712 commented:
oh, I misunderstood the purpose of this PR. The idea I mentioned is to migrate the return type from ```Future``` to ```CompletableFuture```.

What the benefit of reimplementing the ```KafkaFutureImpl``` by ```CompletableFuture``` if we eventually change the return type to ```CompletableFuture```?

dajac commented:
@tombentley Thanks for the PR. I wonder if `KafkaFuture` could implement `CompletionStage` and thus could be directly used with Java Future. Did we consider this?

chia7712 commented:
```KafkaFuture``` is marked as ```Evolving``` so maybe we can open KIP to make it compatible to ```CompletionStage```?

tombentley commented:
@dajac there are at least two reasons why that's problematic:

1. As described by @andrasbeni in his [previous PR](https://github.com/apache/kafka/pull/5131#issuecomment-397024119)
    > The problem with implementing `CompletionStage` is `whenComplete`. Right now user can write
`new KafkaFutureImpl<String>().thenApply(String::length)`. The lambda expression will be compiled to a `KafkaFuture.BiConsumer`. If we implement `CompletionStage`, there will be another `whenComplete` in the same class that takes `java.util.function.BiConsumer`, another SAM interface, as its parameter. That's why compiler will report ambiguity and will fail to compile above code.

2. The exception handling for some methods are just not compatible. For example, the contract of `CompletionStage` says the return value of `CompletionStage#thenApply` (and `whenComplete`) should fail with `CompletionException`:

    >  In all other cases, if a stage's computation terminates abruptly
    >  with an (unchecked) exception or error, then all dependent stages
    >  requiring its completion complete exceptionally as well, with a
    >  {@link CompletionException} holding the exception as its cause.

    But `KafkaFuture`'s `thenApply()` returns a `KafkaFuture` which will fail with `ExecutionException` (because `CompletionException` wasn't in Java 8 which Kafka supported at the time).

(FWIW there's also a mismatch with `getNow` as it works in `CompleteableFuture` and `KafkaFutures` though that's not a method declared on `CompletionStage`).

So trying to make `KafkaFuture` implement `CompleteableFuture` risks breaking users error handling code, possibly in quite subtle ways. Changes to make APIs return `CompletionStage` or `CompletableFuture` rather than `KafkaFuture` are neither source nor binary compatible and users would could have to change their error handling. So I think the only completely _safe_ option is to expose a `toCompletionStage()` method which returns the underlying `CompletableFuture`.

chia7712 commented:
> So I think the only completely safe option is to expose a toCompletionStage() method which returns the underlying CompletableFuture.

This is a good reason of reimplementing the ```KafkaFutureImpl``` by ```CompletableFuture``` :)

tombentley commented:
@dajac, @chia7712 I realise now that I was wrong about 2. It's not a problem because `KafkaFuture.whenComplete` and `thenApply` are using `KafkaFuture.BiConsumer` and `KafkaFuture.Function`, not the ones from `java.util.function`, and so they're distinct from the ones inherited from `CompletionStage`. It would mean subtly different behaviour between the overloaded methods, which isn't ideal, but it is possible. 

That leaves 1, which is a source incompatibility, but not a binary incompatibility. Perhaps, with the `@InterfaceStability.Evolving` that's enough to justify implementing `CompletionStage`, wdyt?

ijuma commented:
`@InterfaceStability.Evolving` has been there for so many releases when it comes to `Admin/AdminClient` that I don't think we can rely on it.

With regards to the method overloads, that would mean that lambdas won't work without casting, right?

tombentley commented:
Casting would be required, yes.

ijuma commented:
I think it would be worth agreeing on the end state. Do we want to have `KafkaFuture` long-term? If not, how do we go about removing its usage? A somewhat extreme approach would be to introduce a new interface `Admin2` (but with a better name) and use `CompletionStage` for all the methods. And eventually deprecate `Admin` and finally remove it.

dajac commented:
@ijuma I was actually thinking about the same. The new or the old could likely be a thin facade of the other. Another approach would be to duplicate all the methods in the *Result classes but that does not sound good to me.

I think that it would be nice if we could eventually remove `KafkaFuture` and use `CompletionStage` instead. That would be nice from a user of the admin api perspective at least. That would also be inline with the producer api KIP-706 lands in the future.

tombentley commented:
Note that `CompletionStage` on its own does not define `get`, `getNow`, which are vital methods for people wanting to use the client in a sync fashion, so I think the comparison is between `Admin2` API which used `CompletableFuture` and `Admin` using a `KafkaFuture` that implemented `CompletionStage`. I suppose you could use `toCompletableFuture`, but then it's less of a drop-in replacement.

A `KafkaFuture` that implemented `CompletionStage` adds very little additional value over `CompletableFuture` in the long term, _except_ the value of backwards compatibility. Anything like `Admin2`:

* would require quite a lot of work by the project (e.g. doubled test surface), albeit not a burden to be carried forever, and maybe the thin facade idea means it wouldn't really be so bad
* incurs a cost to users of discovering and switching to the new API, which offers nothing they couldn't do with a `KafkaFuture` that implemented `CompletionStage`.

My gut feeling is that if the only reason for an `Admin2` was to get rid of `KafkaFuture` then there doesn't seem to be enough of a benefit to justify it. If there were other things which could be cleaned up at the same time then perhaps that would shift the balance.

ijuma commented:
It's a good thing that `CompletionStage` doesn't expose those methods directly, in my opinion. As you said, you can call `toCompletedFuture` to get them if you really want them.

ijuma commented:
> My gut feeling is that if the only reason for an Admin2 was to get rid of KafkaFuture then there doesn't seem to be enough of a benefit to justify it. If there were other things which could be cleaned up at the same time then perhaps that would shift the balance.

I think the challenge is that there is no good way to retrofit `KafkaFuture`. I think the only viable approach I've seen so far is to add the `toCompletableFuture`/`toCompletionStage` methods.

tombentley commented:
> It's a good thing that CompletionStage doesn't expose those methods directly, in my opinion. As you said, you can call toCompletedFuture to get them if you really want them.

It makes perfect sense for the use cases targeted by that API, where random blocking in an async pipeline is to be avoided. But many people who use the Admin client want to do so in a synchronous style. They already have to invoke `KafkaFuture.get` to do that, having to do `toCompletableFuture().get()` makes the API less discoverable and just slightly more painful to those who want sync.  They likely wouldn't see a switch to plain CompletionStage as an improvement. 

> I think the challenge is that there is no good way to retrofit KafkaFuture. I think the only viable approach I've seen so far is to add the toCompletableFuture/toCompletionStage methods.

It seems to be the only completely backwards compatible thing we can do.

ijuma commented:
Another relevant item for consideration is that project Loom indicates that the Java libraries may try to promote async without async APIs.

tombentley commented:
@ijuma I'm not quite sure what you're proposing. That some `Admin2` might not even need to use any kind of Future/CompletionStage API? I don't think that stacks up as an argument for not trying to do something to do something about/improve KafkaFuture today:

* Loom may or may not ultimately be successful in persuading people to do sync programming in virtual threads. Until that's been proven, having an Admin client that supports async is valuable.
* when will virtual threads actually make it into the JDK? My best guess is it won't be production ready for around 18 months in the best case (maybe you have better info than me though?) Even if it was here today people would still be using older JREs for years to come, and some of those people would want to use an Admin client async and use CompletionStage.

Or do I misunderstand the point you're trying to make?

ijuma commented:
@tombentley I didn't suggest anything. :) Maybe Loom means we should avoid Admin2 and go with an incremental path first (maybe something like what you're suggesting here).

ijuma commented:
Also worth thinking about what we would do for new admin methods.

tombentley commented:
Incremental first would be my preference. It doesn't preclude doing something else in the future.

If we improved `KafkaFuture` then I would argue that new admin methods should still use it, for consistency. Allowing a mix of KafkaFuture and CompletionStage just makes the API inconsistent, makes Admin2 feel more necessary than it needs to be, and doesn't add any additional functionality.

tombentley committed:
KIP-707: Add KafkaFuture.toCompletionStage

tombentley committed:
Fix CCE

tombentley committed:
Check JVM version

tombentley commented:
CI faliures are unrelated. @ijuma and/or @chia7712 could one of you review this please?

tombentley commented:
@dajac @ijuma please could you review this, it's the implementation for KIP-707.

tombentley commented:
Test failure is unrelated. @dajac @cmccabe @mimaison please could one of you review?

mimaison commented on clients/src/test/java/org/apache/kafka/common/KafkaFutureTest.java:
```diff
@@ -17,68 +17,261 @@
-        assertEquals("We require more vespene gas", e.getCause().getMessage());
+        assertTrue(futureFail.completeExceptionally(new RuntimeException("We require more vespene gas")));
+        assertIsFailed(futureFail);
+        assertFalse(futureFail.completeExceptionally(new RuntimeException("We require more minerals")));
```
Good to see we're staying on theme!

tombentley commented:
@dajac @chia7712 (and let me try @rajinisivaram on the off-chance), I'd like to get this into Kafka 3.0, but this hasn't seen any review yet despite repeated requests. Any chance we could squeeze it in?

kkonstantine commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -27,217 +28,107 @@
-            waiter.accept(null, newException);
-        }
-        return true;
+        // CF#get() always wraps the _cause_ of a CompletionException in EE (which KF does not)
```
Not sure what EE is here.

tombentley replied:
Expanded all the class abbreviations.

kkonstantine commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -27,217 +28,107 @@
-        KafkaFutureImpl<R> futureImpl = (KafkaFutureImpl<R>) future;
-        futureImpl.addWaiter(new Applicant<>(function, this));
+    private <U> KafkaCompletableFuture<U> toKafkaCompletableFuture(CompletableFuture<U> tCompletableFuture) {
+        KafkaCompletableFuture<U> res;
```
the fact that two variables are used here must be a remainder of an earlier iteration. 
```suggestion
        KafkaCompletableFuture<U> result;
```
can be declared here at the top level and we can remove the inner declaration.

kkonstantine commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -27,217 +28,107 @@
-        KafkaFutureImpl<R> future = new KafkaFutureImpl<>();
-        addWaiter(new Applicant<>(function, future));
-        return future;
+        CompletableFuture<R> completableFuture = this.completableFuture.thenApply((java.util.function.Function<? super T, ? extends R>) value -> {
```
Not a big fan of variable shadowing, which below you avoid by calling the local variable `tCompletableFuture`. Probably good idea to apply the naming here to and remove `this.`

tombentley replied:
Not a fan of the name `tCompletableFuture` either, so I used `appliedFuture` here and made the following method that used `tCompletableFuture` static to avoid shadowing there.

kkonstantine commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -267,50 +173,77 @@ public T get() throws InterruptedException, ExecutionException {
     public synchronized boolean isCancelled() {
-        return exception instanceof CancellationException;
+        if (isDependant) {
+            Throwable exception;
```
I believe just returning from the catch branch is fine. No need for this variable. (you return from any path)

kkonstantine commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -267,50 +173,77 @@ public T get() throws InterruptedException, ExecutionException {
     public String toString() {
-        return String.format("KafkaFuture{value=%s,exception=%s,done=%b}", value, exception, done);
+        T value = null;
+        Throwable ex = null;
```
the naming used above seems better here
```suggestion
        Throwable exception = null;
```

kkonstantine reviewed (commented):
Took a pass and left a few minor comments. 
Looks good to me overall and I could approve if we address asap. Not sure if others have any reservations w/r/t this implementation. If not, it'd be nice to include in 3.0 given that KIP-707 has been adopted. The test coverage seems sufficient.

tombentley committed:
Review comments

dajac commented on clients/src/main/java/org/apache/kafka/common/KafkaFuture.java:
```diff
@@ -98,14 +80,45 @@ private void maybeComplete() {
+        KafkaFutureImpl<Void> result = new KafkaFutureImpl<>();
+        CompletableFuture.allOf(Arrays.stream(futures)
+                .map(kafkaFuture -> {
+                    // Safe since KafkaFuture's only subclass is KafkaFuture for which toCompletionStage()
+                    // always return a CF.
```
nit:  I would replace `CF` by `CompletableFuture`.

dajac commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -27,217 +28,107 @@
-        KafkaFutureImpl<R> future = new KafkaFutureImpl<>();
-        addWaiter(new Applicant<>(function, future));
-        return future;
+        CompletableFuture<R> completableFuture = this.completableFuture.thenApply((java.util.function.Function<? super T, ? extends R>) value -> {
```
nit: Do we really need to specify `(java.util.function.Function<? super T, ? extends R>)` here?

dajac commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -27,217 +28,107 @@
+                return function.apply(value);
+            } catch (Throwable t) {
+                if (t instanceof CompletionException) {
+                    throw new CompletionException(t);
```
I am not sure that I follow why we wrap it here. Could you elaborate?

tombentley replied:
Again it comes down to maintaining the historical behaviour of `thenApply` when the supplied `Function` threw `CompletionException`.  Basically `CompletableFuture` does some unwrapping of `CompletionException` in dependent futures which is incompatible with what `KafkaFuture` did. We can't prevent the unwrapping, but since such wrapping only goes one level deep in the cause chain we can restore the `KafkaFuture` behaviour by inserting a bogus `CompletionException` for `CompletableFuture` to remove. 

I've added a comment and a test for this case.

dajac commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -27,217 +28,107 @@
+    private <U> KafkaCompletableFuture<U> toKafkaCompletableFuture(CompletableFuture<U> tCompletableFuture) {
+        KafkaCompletableFuture<U> res;
+        if (tCompletableFuture instanceof KafkaCompletableFuture) {
+            res = (KafkaCompletableFuture<U>) tCompletableFuture;
```
nit: We could omit `res` and return directly in the two places below.

tombentley replied:
The lambda passed to `whenComplete` needs to complete the `KafkaCompletableFuture`, so needs a captured variable, but I'll declare `result` in the `else` branch only.

dajac commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -27,217 +28,107 @@
-            waiter.accept(null, newException);
-        }
-        return true;
+        // CF#get() always wraps the _cause_ of a CompletionException in EE (which KF does not)
```
nit: I would rather use the full name instead of using acronyms.

dajac commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -247,17 +138,32 @@ public boolean completeExceptionally(Throwable newException) {
+        return completableFuture.cancel(mayInterruptIfRunning);
+    }
+
+    // We need to deal with differences between KF's historic API and the API of CF:
+    // CF#get() does not wrap CancellationException in ExecutionException (nor does KF).
+    // CF#get() always wraps the _cause_ of a CompletionException in ExecutionException (which KF does not).
+    //
+    // The semantics for KafkaFuture are that all exceptional completions of the future (via #completeExceptionally() or exceptions from dependants)
+    // manifest as ExecutionException, as observed via both get() and getNow().
```
nit: Could we turn this into a javadoc or perhaps move it into the method?

dajac commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -247,17 +138,32 @@ public boolean completeExceptionally(Throwable newException) {
+        try {
+            return completableFuture.get();
+        } catch (ExecutionException e) {
+            maybeRewrapAndThrow(e.getCause());
+            throw e;
```
nit: Would it make sense to move `throw e` into `maybeRewrapAndThrow` to let `maybeRewrapAndThrow` throw in both cases? More generally, I wonder if we could handle all the case in `maybeRewrapAndThrow` and use it everywhere.

tombentley replied:
I agree it's a bit messy, but I don't think that works. The call site in `getNow()` differs from the others:

1. It's catching `CompletionException`, not `ExecutionException`
2. In the case there `maybeRewrapAndThrow` didn't throw it needs to wrap the cause in an `ExceptionException`, unlike the other call sites, to conform with the historical API of KafkaFuture. 

The first would require `maybeRewrapAndThrow()` to have an `Exception`-typed parameter, which, being a checked exception, means we'd need either to declare `throws Exception`, or we'd need to type cast within `maybeRewrapAndThrow()` and be left with an `else` case to deal with. 

The second is difficult to abstract over nicely for the other call sites: We'd need to pass a boolean to indicate whether the `ExceptionException` was needed.

dajac replied:
How about defining two helper methods, one for each cases?
* `private void maybeRewrapAndThrow(ExecutionException exception)`; and
* `private void maybeRewrapAndThrow(CompletionException exception)`

dajac replied:
@tombentley What do you think about this? Would it help to consolidate the logic?

tombentley replied:
It doesn't seem to help with other places where we catch and rethrow because they catch unchecked exceptions `Error` and `RuntimeException` whose only supertype is `Throwable`, so we end up in a painful place due to Java's lack of abstraction for "unchecked exception".

For the existing call sites the `void` return type you proposed doesn't work with javac's definite return analysis so I have to declare `private <T> T rewrapAndThrow(CompletionException exception)` and `private <T> T rewrapAndThrow(ExecutionException exception)`. The method always throws, so the return type of `T` is just there to appease `javac` via a `return rewrapAndThrow(...)` at the call site. 

I can do this is you want, but I feel it's making the code less clear. The `rewrapAndThrow(CompletionException exception)` only has a single caller, so it's of little benefit.

I guess the other alternative would be to inline the method and just comment each site as to what's going on.

dajac replied:
I don't like the fact that we throw in two places... Could we at least make the name of `maybeRewrapAndThrow` a bit more explicit? It is only about `CancellationException` in the end so we could name it `maybeThrowCancellationException` or something like this. Moreover, the method does not really "rewrap" anything, right? It just checks the type and throw it.

tombentley replied:
There don't seem to be any good options here.

We could throw from a single method if you're willing to have the method like this

```java
    private <T, E extends Exception> T throwUnwrappingCancellationException(E exception) throws E {
        if (exception.getCause() instanceof CancellationException) {
            throw (CancellationException) exception.getCause();
        }
        throw exception;
    }
```

and the call sites to look like this:

```java
return throwUnwrappingCancellationException(e);
```

 The `return <methodThatAlwaysThrows>` is nasty, and I can't think of a good method name. Personally I don't think this is better than naming the current method `maybeThrowCancellationException` as you suggest, and living with the throw in two places, so I've gone with that.

dajac commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -267,50 +173,77 @@ public T get() throws InterruptedException, ExecutionException {
+            } catch (Exception e) {
+                exception = e;
+            }
+            return exception instanceof CompletionException
+                    && exception.getCause() instanceof CancellationException;
```
I am curious here. Would it be possible to let the `CancellationException` propagate from the upstream future to this one when the upstream future is cancelled?

tombentley replied:
The historical API of `KafkaFuture` on cancellation of an upstream future (e.g. `thenApply()`) was for both futures to return true from `isDone()`, `isCancelled()` and `isCompletedExceptionally()`, and for both `getNow` and `get()` to throw a `CancellationException`. 

If the new implementation of `isCancelled()` just does `return completableFuture.isCancelled();` then for a dependent future `isCancelled()` would return false because it's exception is a `CompletionException` wrapping a `CancellationException` and `CompleteableFuture#isCancelled()` just checks for the exception being `CancellationException`. So we need to do something different for cancellation of dependent futures. 

Because `KafkaFutureImpl.thenApply()` doesn't itself keep track of waiters any more it's not possible to do anything clever in the implementation of `cancel()`, so we have to do something in `isCancelled()`.

So to (hopefully) answer your question: It does propagate already, but the APIs differ and what we're doing here is to maintain the existing semantics of `KafkaFuture`.

dajac replied:
Oh.. I see. Thanks for the clarification. It would be great if you could add a comment which explains this because this is not obvious.

tombentley replied:
Added. Let me know whether it makes sense to you. Also: this is covered by the test which passes with the old and new implementations.

dajac reviewed (commented):
@tombentley I am really sorry for the delay on this one. I finally had the time to look at it. Overall, it looks pretty good to me. I left few comments and questions.

tombentley committed:
Review comments @dajac

tombentley committed:
test

tombentley committed:
Clarifying comment

tombentley committed:
Improve test

tombentley commented:
@kkonstantine @dajac thanks for the reviews, I really appreciate it. I think I've addressed all the comments and hopefully answered your questions, if you'd like to take another look?

tombentley committed:
comment

tombentley committed:
rename method

kkonstantine reviewed (approved):
Thanks @tombentley 
LGTM
It'd be best to have @dajac's approval too.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaCompletableFuture.java:
```diff
@@ -0,0 +1,79 @@
+ * Kafka API to a client application.
+ * @param <T> The type of the future value.
+ */
+public class KafkaCompletableFuture<T> extends CompletableFuture<T> {
```
Why not throwing exception when users call `toCompletableFuture`? If users want to use blocking method (ex. `get()`), they ought to call `KafkaFuture#get` instead. What is the benefit if we expose `toCompletableFuture`?

tombentley replied:
The KIP describes it like this:

> Adding toCompletionStage() is sufficient because CompletionStage itself exposes toCompletableFuture(), so anyone who needs an actual CompletableFuture (e.g. for interoperating with 3rd party APIs that require one) can get one. However, CompletableFuture exposes methods for future completion which should not be called by users (only the Admin client should be completing the returned futures), so calling these will be prevented. It is expected that users wanting to block on the completion of the KafkaFuture would use kafkaFuture.get(), rather than calling kafkaFuture.toCompletionStage().toCompletableFuture().get(), so the need to access the CompleteableFuture should be rare.

Essentially we want to allow access to a `CompletableFuture` to allow interop with 3rd part APIs which might require it (e.g. for using the blocking `get` methods) , but not allow user code to complete those futures. The `CompletableFuture` API unhelpfully adds _both_ the blocking `get` methods and the `complete` methods.

chia7712 replied:
> allow interop with 3rd part APIs which might require it

fair enough :)

chia7712 commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaCompletableFuture.java:
```diff
@@ -0,0 +1,79 @@
+ */
+public class KafkaCompletableFuture<T> extends CompletableFuture<T> {
+
+    boolean kafkaComplete(T value) {
```
why it needs prefix "kafka"?

tombentley replied:
The idea is that all Kafka code knows it's using a `KafkaCompletableFuture` and knows to complete it via the `kafka*()` methods. User code which erroneously tries to complete via `complete()` gets the exception thrown.

chia7712 replied:
> The idea is that all Kafka code knows it's using a KafkaCompletableFuture and knows to complete it via the kafka*() methods.

that is a good point. Could you add this comment?

tombentley replied:
Done

chia7712 commented on clients/src/main/java/org/apache/kafka/common/KafkaFuture.java:
```diff
@@ -19,75 +19,57 @@
+ * {@code KafkaFuture} will fail with an {@code ExecutionException}, whereas a {@code CompletionStage} fails
+ * with a {@code CompletionException}.
  */
 @InterfaceStability.Evolving
```
Should we remove this annotation.

KIP-707 has following description.

```
Remove the @InterfaceStability.Evolving annotation on KafkaFuture to reflect the reality that changing this class incompatibly would cause of too much user code to break.
```

chia7712 commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -267,50 +180,82 @@ public T get() throws InterruptedException, ExecutionException {
+        if (isDependant) {
+            // Having isCancelled() for a dependent future just return
+            // CompletableFuture.isCancelled() would break the historical KafkaFuture behaviour because
+            // CompleteableFuture#isCancelled() just checks for the exception being CancellationException
```
typo: CompleteableFuture -> CompletableFuture

chia7712 commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -267,50 +180,82 @@ public T get() throws InterruptedException, ExecutionException {
+            // CompletableFuture.isCancelled() would break the historical KafkaFuture behaviour because
+            // CompleteableFuture#isCancelled() just checks for the exception being CancellationException
+            // whereas it will be a CompletionException wrapping a CancellationException
+            // due needing to compensate for CompleteableFuture's CompletionException unwrapping
```
typo: CompleteableFuture -> CompletableFuture

chia7712 commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -267,50 +180,82 @@ public T get() throws InterruptedException, ExecutionException {
-        if (done)
-            return value;
-        return valueIfAbsent;
+    public synchronized T getNow(T valueIfAbsent) throws ExecutionException {
```
Does it need to remove `InterruptedException` from other get methods also? For another, why we don't remove `InterruptedException` from `KafkaFuture#getNow`?

tombentley replied:
> Does it need to remove InterruptedException from other get methods also?

`get()` and `get(long, TimeUnit)` can still throw `InterruptedException` due to the blocking behaviour (the `completableFuture.get` that we delegate to throws `InterruptedException`).

> For another, why we don't remove InterruptedException from KafkaFuture#getNow?

Good spot!  `KafkaFuture#getNow` should never have declared that and it's annoying for callers to have to deal with it. But this wasn't described in the KIP and wouldn't be a source compatible change (existing code with a `catch (InterruptedException)` would get a compile error). 

I'd be happy to remove it and notify the VOTE thread, WDYT @kkonstantine @dajac ?

dajac replied:
Good question. AK 3.0 is a good opportunity to do such breaking changes so I would be in favour of doing it. Let's see what other think.

chia7712 replied:
> But this wasn't described in the KIP and wouldn't be a source compatible change (existing code with a catch (InterruptedException)

As it can cause compatible change, we should deprecate it and then add a new method (for example: get(T value)) to replace it. This can be discussed in another issue :)

tombentley replied:
I wouldn't call it `get` because it has different semantics to the existing get methods, so I think it should have a distinct name. And I can't think of a succinct alternative to `getNow`. Something like `getOrDefault` doesn't convey they key difference from `get` (the lack of blocking). `getImmediately` is just a bit verbose. 

The API of KafkaFuture has always been a subset of `CompletableFuture`. Introducing a new method would break that. While it would be discoverable while the deprecated method remained to point the way it wouldn't be as discoverable once the old `getNow` was removed.

All deprecation really does is let the programmer defer fixing the problem. But it's most likely not a huge problem to fix (i.e. they've likely caught the exception close to the method call, rather than defining many layers of API with `throws InterruptedException`).

A new method also means their code is then not binary compatible with older client versions. 

So on balance I personally favour just removing the `throws InterruptedException` in 3.0 and letting the programmer clean up the code sooner rather than later. 

There is always the "do nothing" alternative.

ijuma replied:
"do nothing" is probably the right thing here.

tombentley replied:
OK, I've reverted it.

chia7712 replied:
> A new method also means their code is then not binary compatible with older client versions.

yep, the BC could be broken. What I really care is the source compatibility. It seems to me a public API should be source compatible to next major release :)

This patch is good to go and we need more time (rather than vespene gas ... StarCraft is my favor game :) ) to reach consensus.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -247,17 +142,35 @@ public boolean completeExceptionally(Throwable newException) {
      */
     @Override
     public synchronized boolean cancel(boolean mayInterruptIfRunning) {
```
Do we still need `synchronized`? It seems to me `CompletableFuture` can deal with thread issue.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaCompletableFuture.java:
```diff
@@ -0,0 +1,79 @@
+
+    @Override
+    public boolean complete(T value) {
+        throw new UnsupportedOperationException();
```
How about adding some comments to this exception?

chia7712 reviewed (commented):
@tombentley thanks for this good patch. This issue is hard but important. a couple of comments below.

tombentley committed:
Chia-Ping's comments

tombentley commented:
@chia7712 thanks for the review!

tombentley commented:
@dajac @chia7712 I think I've addressed all your comments.

dajac commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -27,217 +28,111 @@
-    }
 
     @Override
     public synchronized boolean complete(T newValue) {
```
Could we also remove `synchronized` here?

dajac commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -267,50 +180,82 @@ public T get() throws InterruptedException, ExecutionException {
      * the given valueIfAbsent.
      */
     @Override
-    public synchronized T getNow(T valueIfAbsent) throws InterruptedException, ExecutionException {
```
Could we also remove `synchronized` here?

tombentley committed:
Remove unneeded synchronized

dajac reviewed (approved):
LGTM, thanks for the patch. It would be great to have +1 from @chia7712 as well.

chia7712 commented on clients/src/test/java/org/apache/kafka/common/KafkaFutureTest.java:
```diff
@@ -17,68 +17,265 @@
+            throw new AssertionError("Unexpected exception", e);
+        }
+        try {
+            future.getNow(null);
```
How about using following style.

```java
        ExecutionException e = assertThrows(ExecutionException.class, () -> future.getNow(null));
        assertEquals(expectedException, e.getCause().getClass());
        assertEquals(expectedMessage, e.getCause().getMessage());
        return e.getCause();
```

chia7712 commented on clients/src/main/java/org/apache/kafka/common/KafkaFuture.java:
```diff
@@ -184,7 +194,7 @@ public abstract T get(long timeout, TimeUnit unit) throws InterruptedException,
      * the given valueIfAbsent.
      */
-    public abstract T getNow(T valueIfAbsent) throws InterruptedException, ExecutionException;
+    public abstract T getNow(T valueIfAbsent) throws ExecutionException;
```
Could you please revert this change as you mentioned that it can cause compatible issue.

chia7712 commented on clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java:
```diff
@@ -27,217 +28,111 @@
-                }
-            }
-        }
+    public KafkaFutureImpl(KafkaCompletableFuture<T> completableFuture) {
```
Is this method still useful?

tombentley replied:
No, but it makes me wonder whether we'd find factory methods for already completed and already failed `KafkaFutures` useful.

chia7712 reviewed (commented):
@tombentley thanks for updating code. a couple of comments below. PTAL

tombentley committed:
assertThrows

tombentley committed:
Revert InterruptedException

tombentley committed:
remove constructor

tombentley commented:
@chia7712 thanks, I've implemented your most recent comments.

tombentley commented:
If no further comments I'll merge this in the next hour or so.

chia7712 reviewed (approved):
LGTM

tombentley commented:
Thanks for the reviews @chia7712 @dajac @kkonstantine
